File: 560-subarray-sum-equals-k/subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        mp = {}
        tot = 0
        count = 0
        for ele in nums:
            tot += ele
            if(tot == k):
                count += 1
            if(tot - k in mp):
                count += mp[tot-k]
            if(tot in mp):
                mp[tot] += 1
            else:
                mp[tot] = 1

        return count

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Long, Integer> map = new HashMap<>();
        long sum = 0;
        int ans = 0;
        for(int ele : nums) {
            sum += ele;
            if(sum == k) ans++;
            if(map.containsKey(sum - k)) {
                ans += map.get(sum - k);
            }
            map.put(sum, map.getOrDefault(sum, 0) + 1);
        }
        return ans;
    }
}